### PR TITLE
Add Toggle implementation

### PR DESCRIPTION
**Pull request by j-f1**
I think it’s fine to merge this without the toggle style since that’s not super common on the web.

### CODE REVIEW

File: Sources/TokamakDOM/Styles/ToggleStyle.swift [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        },
+      ])
+      configuration.label
+    }
+  }
+}
+
+// FIXME: implement
+// public struct SwitchToggleStyle: ToggleStyle {
+//    public func makeBody(configuration: ToggleStyleConfiguration) -> HTML<TupleView<(HTML<EmptyView>, AnyView)>> {
```
Comment (ie-ahm-robox) [inline Sources/TokamakDOM/Styles/ToggleStyle.swift:46]: - :no_entry_sign: Line should be 100 characters or less: currently 116 characters (`line_length`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Styles/ToggleStyle.swift#L17** - Line should be 100 characters or less: currently 115 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Styles/ToggleStyle.swift#L54** - Line should be 100 characters or less: currently 118 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Toggle.swift#L28** - Line should be 100 characters or less: currently 101 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/DOMRenderer.swift#L62** - Line should be 100 characters or less: currently 127 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/Shapes/Path.swift#L51** - Line should be 100 characters or less: currently 115 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/Shapes/Path.swift#L57** - Line should be 100 characters or less: currently 113 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/Styles/ToggleStyle.swift#L27** - Line should be 100 characters or less: currently 112 characters (`line_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against 7666e45e238fa29c6b16db88c960e774cc709ea2
</p>

File: Sources/TokamakDOM/Shapes/Path.swift [hunk 2]
```diff
@@ -50,7 +50,7 @@ extension Path: ViewDeferredToRenderer {
         "rx": "\(roundedRect.cornerSize.width)",
         "ry": "\(roundedRect.style == .continuous ? roundedRect.cornerSize.width : roundedRect.cornerSize.height)",
```
Comment (ie-ahm-robox) [inline Sources/TokamakDOM/Shapes/Path.swift:51]: - :no_entry_sign: Line should be 100 characters or less: currently 115 characters (`line_length`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/Shapes/Path.swift#L57** - Line should be 100 characters or less: currently 113 characters (`line_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against 219087122284fffa36fc8c65a6e3604a8000a2bb
</p>

File: Sources/TokamakDOM/Shapes/Path.swift [hunk 3]
```diff
@@ -44,17 +44,21 @@ extension Path: ViewDeferredToRenderer {
       return AnyView(HTML("ellipse", ["cx": "50%", "cy": "50%", "rx": "50%", "ry": "50%"]
           .merging(stroke, uniquingKeysWith: uniqueKeys)))
     case let .roundedRect(roundedRect):
+      let ry = roundedRect.style == .continuous
+        ? roundedRect.cornerSize.width
+        : roundedRect.cornerSize.height
       return AnyView(HTML("rect", [
         "width": "\(roundedRect.rect.size.width)",
```
Comment (ie-ahm-robox) [inline Sources/TokamakDOM/Shapes/Path.swift:51]: - :no_entry_sign: Line should be 100 characters or less: currently 115 characters (`line_length`)

File: Sources/TokamakDOM/DOMRenderer.swift [hunk 4]
```diff
@@ -59,16 +59,21 @@ public final class DOMRenderer: Renderer {
 
   public init<V: View>(_ view: V, _ ref: JSObjectRef) {
     rootRef = ref
-    rootRef.style = "display: flex; width: 100%; height: 100%; justify-content: center; align-items: center; overflow: hidden;"
+    rootRef.style = "display: flex; width: 100%; height: 100%;"
+      + "justify-content: center; align-items: center; overflow: hidden;"
```
Comment (MaxDesiatov) [inline Sources/TokamakDOM/DOMRenderer.swift:63]: Multiline strings are easier to read, and this also would fix the compilation error 🙂

```suggestion
    rootRef.style = """
    display: flex; width: 100%; height: 100%;
    justify-content: center; align-items: center; overflow: hidden;
    """
```

File: Sources/TokamakCore/Styles/ToggleStyle.swift [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+  func makeBody(configuration: Self.Configuration) -> Self.Body
+
+  typealias Configuration = ToggleStyleConfiguration
+}
+
+public struct AnyToggleStyle: ToggleStyle {
+  public typealias Body = AnyView
+
+  private var bodyBuild: (ToggleStyleConfiguration) -> AnyView
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Styles/ToggleStyle.swift:42]: **nitpick**

```suggestion
  private var bodyBuild: (ToggleStyleConfiguration) -> AnyView
```

**Comment by j-f1** [comment]
Weird issue: when clicking the “Check me!” checkbox, I get this error: “Fatal error: ToggleStyleKey must have a renderer-provided default value.” Somehow, the default environment values are not being provided.

File: Sources/TokamakCore/Styles/ToggleStyle.swift [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    }
+  }
+
+  public func makeBody(configuration: ToggleStyleConfiguration) -> AnyView {
+    bodyClosure(configuration)
+  }
+}
+
+public enum ToggleStyleKey: EnvironmentKey {
+  public static var defaultValue: AnyToggleStyle {
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Styles/ToggleStyle.swift:56]: This is the single `defaultValue` provided for this key. Thus `fatalError` is what we get, which needs to change for it to work.

Comment (j-f1) [reply Sources/TokamakCore/Styles/ToggleStyle.swift:56]: Yes, but the default environment includes a value for the key:

```swift
var environment = EnvironmentValues()
environment[ToggleStyleKey] = AnyToggleStyle(DefaultToggleStyle())
```

**Comment by MaxDesiatov** [comment]
Ah, this may be related to the bug I've seen where `mount` functions do inject the environment, but `update` don't. Having a look now...

**Comment by MaxDesiatov** [comment]
Sorry for the delay, I've stumbled upon a different (but kinda related) bug that I'm finding hard to fix without proper stacktraces, so I'm thinking that the entrypoint/polyfills code in `carton` needs to be updated to support stacktraces, also hoping to fix https://github.com/swiftwasm/carton/issues/25 at the same time with the update. Yet another rabbit hole for me `¯\_(ツ)_/¯`

**Comment by j-f1** [comment]
No worries! I’m fine waiting.

**Comment by carson-katri** [comment]
This may be unblocked with the changes to environment in #170

**Comment by j-f1** [comment]
Thanks, @carson-katri! The new version seems to be working well, with the caveat that the “I’m always checked!” checkbox can still be unchecked.

**Comment by j-f1** [comment]
Done!

File: Sources/TokamakCore/Styles/ToggleStyle.swift [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+public protocol ToggleStyle {
+  associatedtype Body: View
+
+  func makeBody(configuration: Self.Configuration) -> Self.Body
+
+  typealias Configuration = ToggleStyleConfiguration
+}
+
+public struct AnyToggleStyle: ToggleStyle {
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Styles/ToggleStyle.swift:39]: **nitpick**: this isn't a part of the public SwiftUI API, so under the "package internal" rules I think it should be underscored

```suggestion
public struct _AnyToggleStyle: ToggleStyle {
```

File: Sources/TokamakDOM/Styles/ListStyle.swift [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+//  Created by Jed Fox on 06/30/2020.
+//
+
+import TokamakCore
+
+public typealias DefaultListStyle = TokamakCore.DefaultListStyle
+public typealias PlainListStyle = TokamakCore.PlainListStyle
+public typealias InsetListStyle = TokamakCore.InsetListStyle
+public typealias GroupedListStyle = TokamakCore.GroupedListStyle
+public typealias InsetGroupedListStyle = TokamakCore.InsetGroupedListStyle
```
Comment (MaxDesiatov) [inline Sources/TokamakDOM/Styles/ListStyle.swift:24]: These re-exports should be moved to `Core.swift` probably. I'm also getting `invalid redeclaration` build errors, which seem to be reproducible on CI.

File: Sources/TokamakDOM/Views/Toggle.swift [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//  Created by Jed Fox on 07/04/2020.
+//
+
+import TokamakCore
+
+public typealias Toggle = TokamakCore.Toggle
```
Comment (MaxDesiatov) [inline Sources/TokamakDOM/Views/Toggle.swift:20]: This also belongs to `Core.swift` I guess

File: Sources/TokamakDemo/ListDemo.swift [hunk 10]
```diff
@@ -44,6 +44,6 @@ public struct ListDemo: View {
           Text("Item: \($0 + 1)")
         }
       }
-    }
+    }.frame(width: 400, height: 300)
```
Comment (MaxDesiatov) [inline Sources/TokamakDemo/ListDemo.swift:47]: Not sure if this `frame` change is related to this PR?

Comment (MaxDesiatov) [reply Sources/TokamakDemo/ListDemo.swift:47]: I see, it collapses the list 👍

File: Sources/TokamakDemo/ListDemo.swift [hunk 11]
```diff
@@ -18,17 +18,6 @@
 import TokamakShim
 
 public struct ListDemo: View {
-  let fs: [File] = [
```
Comment (MaxDesiatov) [inline Sources/TokamakDemo/ListDemo.swift:21]: was this removed intentionally?

Comment (j-f1) [reply Sources/TokamakDemo/ListDemo.swift:21]: Yep! It’s not used here.

File: Sources/TokamakDemo/ListDemo.swift (hunk 10 above)
Comment (j-f1) [reply Sources/TokamakDemo/ListDemo.swift]: I reverted it anyway

Comment (MaxDesiatov) [reply Sources/TokamakDemo/ListDemo.swift]: I guess it won't be needed as much when it's more organized after the navigation view PR is merged...

**Pull request merged**